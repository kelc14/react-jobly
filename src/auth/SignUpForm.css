.SignUpForm {
  width: 50vw;
  background-color: white;
  border: 1px solid black;
  border-radius: 10px;
  margin: 0 auto;
  margin-top: 10%;
  padding: 20px;
  overflow: hidden;
}
.SignUpForm-heading {
  color: goldenrod;
}
.SignUpForm-form {
  text-align: left;
}
.SignUpForm-label {
  display: block;
  padding-top: 10px;
  font-size: 1rem;
  font-weight: 500;
}
.SignUpForm-input {
  width: 100%;
  padding: 12px 20px;
  margin: 8px 0;
  box-sizing: border-box;
}
.SignUpForm-button {
  background-color: goldenrod;
  border: none;
  border-radius: 10px;
  color: white;
  padding: 16px 32px;
  font-size: 1rem;
  margin: 4px 2px;
  cursor: pointer;
}
