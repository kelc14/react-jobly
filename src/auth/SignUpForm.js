import React, { useState } from "react";
import "./SignUpForm.css";

const SignUpForm = ({ loginUser }) => {
  const INITIAL_STATE = {
    username: "",
    password: "",
    firstName: "",
    lastName: "",
    email: "",
  };
  const [formData, setFormData] = useState(INITIAL_STATE);

  /** Send {USERNAME, PASSWORD} to API to check if logged in and provide feedback
   *    & clear form. */

  const handleSubmit = (evt) => {
    evt.preventDefault();
    loginUser();
    setFormData(INITIAL_STATE);
  };

  /** Update local state w/curr state of input elem */

  const handleChange = (evt) => {
    const { name, value } = evt.target;
    setFormData((fData) => ({
      ...fData,
      [name]: value,
    }));
  };

  /** Render:
   *
   */
  return (
    <div className="SignUpForm" onSubmit={handleSubmit}>
      <h2 className="SignUpForm-heading">Sign Up</h2>
      <form className="SignUpForm-form">
        <label htmlFor="username" className="SignUpForm-label">
          Username:
        </label>
        <input
          id="username"
          type="text"
          name="username"
          onChange={handleChange}
          value={formData.username}
          className="SignUpForm-input"
        />

        <label htmlFor="password" className="SignUpForm-label">
          Password:
        </label>
        <input
          id="password"
          type="password"
          name="password"
          onChange={handleChange}
          value={formData.password}
          className="SignUpForm-input"
        />

        <label htmlFor="firstName" className="SignUpForm-label">
          First Name:
        </label>
        <input
          id="firstName"
          type="text"
          name="firstName"
          onChange={handleChange}
          value={formData.firstName}
          className="SignUpForm-input"
        />
        <label htmlFor="lastName" className="SignUpForm-label">
          Last Name:
        </label>
        <input
          id="lastName"
          type="text"
          name="lastName"
          onChange={handleChange}
          value={formData.lastName}
          className="SignUpForm-input"
        />
        <label htmlFor="email" className="SignUpForm-label">
          Email Address:
        </label>
        <input
          id="email"
          type="email"
          name="email"
          onChange={handleChange}
          value={formData.email}
          className="SignUpForm-input"
        />

        <button className="SignUpForm-button">Sign Up</button>
      </form>
    </div>
  );
};

export default SignUpForm;
