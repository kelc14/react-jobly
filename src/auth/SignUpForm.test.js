import React from "react";
import { render, screen, waitFor, act } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MemoryRouter } from "react-router-dom";
import SignUpForm from "./SignUpForm";
import JoblyApi from "../api/api";

/**  SMOKE TEST */
it("renders without crashing", function () {
  render(
    <MemoryRouter>
      <SignUpForm />
    </MemoryRouter>
  );
});

/**  SNAPSHOT TEST */
it("renders without crashing", function () {
  const { asFragment } = render(
    <MemoryRouter>
      <SignUpForm />
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

// doesn't work:

// /** SIGNUP FORM e2e */
describe("SignUpForm", () => {
  it("signs up a user and calls loginUser", async () => {
    const loginUserMock = jest.fn();
    render(
      <MemoryRouter>
        <SignUpForm loginUser={loginUserMock} />
      </MemoryRouter>
    );

    const usernameInput = screen.getByLabelText("Username:");
    const passwordInput = screen.getByLabelText("Password:");
    const firstNameInput = screen.getByLabelText("First Name:");
    const lastNameInput = screen.getByLabelText("Last Name:");
    const emailAddressInput = screen.getByLabelText("Email Address:");
    const signUpBtn = screen.getByText("Sign Up");

    act(() => {
      userEvent.type(usernameInput, "testuser1234");
      userEvent.type(passwordInput, "password");
      userEvent.type(firstNameInput, "Test First");
      userEvent.type(lastNameInput, "Test Last");
      userEvent.type(emailAddressInput, "test@example.com");
    });

    // Mock the userSignup function
    const userSignupMock = jest.fn().mockResolvedValue("mockToken");
    jest.spyOn(JoblyApi, "userSignup").mockImplementation(userSignupMock);

    act(() => {
      userEvent.click(signUpBtn);
    });

    // Verify that the userSignup function was called with the expected data
    expect(userSignupMock).toHaveBeenCalledWith({
      username: "testuser1234",
      password: "password",
      firstName: "Test First",
      lastName: "Test Last",
      email: "test@example.com",
    });

    // Verify that the loginUser function was called with the correct arguments
    await waitFor(() => {
      expect(loginUserMock).toHaveBeenCalledWith("testuser1234", "mockToken");
    });
  });
});
