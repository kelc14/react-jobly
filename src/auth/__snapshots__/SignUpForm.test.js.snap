// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="SignUpForm"
  >
    <h2
      class="SignUpForm-heading"
    >
      Sign Up For Jobly Account
    </h2>
    <form
      class="SignUpForm-form"
    >
      <label
        class="SignUpForm-label"
        for="username"
      >
        Username:
      </label>
      <input
        autocomplete="username"
        class="SignUpForm-input"
        id="username"
        name="username"
        type="text"
        value=""
      />
      <label
        class="SignUpForm-label"
        for="password"
      >
        Password:
      </label>
      <input
        autocomplete="new-password"
        class="SignUpForm-input"
        id="password"
        name="password"
        type="password"
        value=""
      />
      <label
        class="SignUpForm-label"
        for="firstName"
      >
        First Name:
      </label>
      <input
        class="SignUpForm-input"
        id="firstName"
        name="firstName"
        type="text"
        value=""
      />
      <label
        class="SignUpForm-label"
        for="lastName"
      >
        Last Name:
      </label>
      <input
        class="SignUpForm-input"
        id="lastName"
        name="lastName"
        type="text"
        value=""
      />
      <label
        class="SignUpForm-label"
        for="email"
      >
        Email Address:
      </label>
      <input
        class="SignUpForm-input"
        id="email"
        name="email"
        type="email"
        value=""
      />
      <button
        class="SignUpForm-button"
      >
        Sign Up
      </button>
    </form>
  </div>
</DocumentFragment>
`;
